pravega/pravega
# Issue 2228: (Segment Store) Admin Tools

**Change log description**
- Built a shell for a command-line interface that can support an arbitrary set of commands a user wants to run.
- Able to load Pravega config from a config file and set custom configs on it
- BookKeeper commands:
    - List all BookKeeperLog summaries
    - List details for a BookKeeperLog
    - Cleanup orphan ledgers (issue #1165)
    - Enabling a `BookKeeperLog` that is currently disabled.
    - Disabling a `BookKeeperLog` that is currently enabled.
- Container commands:
    - Executing a non-invasive `DurableLog` recovery for a specified container
        - Non-invasive means it is performed without the need to fence out Tier1 or otherwise make any sort of modifications to the data in Tier1.
        - `DurableLog` recovery means we are only reading data from Tier1, deserializing `Operations` and adding them a `ReadIndex` (no cache) and validating Metadata. This is the exact same process as in a normal SegmentContainer recovery, but we don't start other sub-services, such as `OperationProcessor` or `StorageWriter`.
- Updated the SelfTester to pause for user input towards the end of the test - this allows a quick debugging of the admin tools, by pointing them to a local cluster that already has data and/or is active

**Purpose of the change**
Fixes #2228.
Fixes #1165.
Fixes #2261.
Fixes #2262. 

**What the code does**
- New package (that currently lives under `SegmentStore/Host`) that enables running arbitrary commands that are registered. In the future, this may be moved to a new package, but I did not want to do that at the moment since I would have had to pull a large number of dependencies into that new package. When the time comes to add more tools, we can execute this move.
    - This can be easily extended to allow new types of commands targeting different components (i.e., Controller, or ZK directly or BK directly). When we will get there, we can make the necessary refactoring to accommodate such new needs. For now, the only target of this is the Segment Store, which is why it lives under it.
- Recovery code inside `DurableLog` has been factored out
- All admin tool access is done via "Debug Wrappers" and no changes to actual functionality has been made to the greatest extent possible.

**How to verify it**
Open a separate terminal, and type:
```
./gradlew test:integration:selftest -Dtarget=OutOfProcess -Dreads=false -Dc=4 -Ds=1 -Dsc=16 -Dws=1000 -Do=1000000 -Dp=100 -Dpp=100 -Dpause=true
```
(make sure `-Dpause=true` is set)

Wait at least until the test started or until you are instructed to press any key to continue (do not press anything).

0. To ease your pain, create a new file, name it `admin.properties` and set its contents to:
```
pravegaservice.containerCount=4
pravegaservice.zkURL=localhost:9000
bookkeeper.bkLedgerPath=/pravega/selftest/bookkeeper/ledgers
bookkeeper.zkMetadataPath=/pravega/selftest/segmentstore/containers
bookkeeper.bkEnsembleSize=1
bookkeeper.bkAckQuorumSize=1
bookkeeper.bkWriteQuorumSize=1
```
1. In a separate terminal, type `./gradlew test:integration:selftest -Dtarget=OutOfProcess -Dreads=false -Dc=4 -Ds=1 -Dsc=16 -Dws=1000 -Do=1000000 -Dp=100 -Dpp=100 -Dpause=true`
    - Wait until you are prompted to press any key to continue
2. In another terminal or in the IDE, run, the `AdminRunner` main program, and pass `-Dpravega.configurationFile={path to file from Step 0}`. If you had run this from a cluster, you'd be pointing to the Pravega config file that is on that cluster.
3. type `help` to see a list of available commands
4. try the following commands:
    - `config list`: You should see the configuration you've set.
    - `bk list`:  You should see all available logs.
    - `bk details 0` (or 1, 2, or 3 - pick from the previous step). You should see details about each BookKeeperLog, including ledger info.
    - `bk cleanup`. This should say there's nothing to clean up. You can change the SelfTester to create arbitrary ledgers, which would show up here.
    - `bk disable 0` - disable log 0 (NOTE: this will cause your running Pravega instance to shutdown container 0 and restart it in offline mode)
    - `bk enable 0` - re-enable log 0 (in a couple of minutes, your Running Pravega instance will resume its normal operations
    - `container recover 0` - attempt local non-intrusive recovery of Container 0. It really doesn't matter if the log is enabled or not; if it's disabled you are guaranteed to get all the latest data, othewise this might not display the latest one (that's how BK works)
